### ISSUE TITLE
Queries with polymorphic types fail at run-time

### ISSUE THREAD

**Issue by jamescheney** [links-lang/links#490]
Thanks to @SimonJF for running into this.

Consider the following (using the tutorial factorials database):
```
links> var db = database "links";
links> var factorials = table "factorials" with (i : Int, f : String) from db;
links> query { for (x <-- factorials) [] };
*** Error: "Assert_failure core/query.ml:156:62"
links> query { for (x <-- factorials) [(a=1)] };
[(a = 1), (a = 1), (a = 1), (a = 1), (a = 1), (a = 1), (a = 1), (a = 1),
 (a = 1), (a = 1), (a = 1), (a = 1), (a = 1), (a = 1), (a = 1), (a = 1),
 (a = 1), (a = 1), (a = 1), (a = 1)] : [(a:Int)]
```

If the query result is an empty list then the assertion failure results from `Query.type_of_expression` trying to guess the query result type and failing because the empty list can have any type.

The/a principled way to handle this would be to retain enough information about the result types of queries (possibly using run-time type information) to do the right thing.

A workaround for now is to avoid writing queries that don't have easy-to-determine types (according to `Query.type_of_expression`).  We should also change the error message to something more enlightening.

**Comment by SimonJF** [comment]
Somewhat worryingly, I've found that this can also manifest itself at runtime, when the result of a query returns the empty list. I haven't properly shrunk this example yet, but consider the following fragment of a query:

```
 [..]
              for (pp <-- Tables.pathophysiology)
                where (pp.disease_id == diseaseId)
                for (o <-- Tables.object)
                  where (o.object_id == pp.object_id)
                    [(id=o.object_id, name=o.name, comments=pp.comments,
                      role=pp.role, drugComments=pp.drugs, sideEffects=pp.side_effects,
                      therapeuticUse=pp.use,
                      references =
                        for (pp_ref <-- Tables.pathophysiology_refs)
                          where (pp_ref.pathophysiology_id == pp.pathophysiology_id)
                          [pp_ref.reference_id])]
```

I've found that if the result of the query determining `references` returns an empty result set, then the whole program errors with the above assert failure. This seems like it only happens with nested queries (several levels deep, usually).

Putting the result in a record type doesn't seem to help, either.

**Comment by SimonJF** [comment]
So I hacked around this:
https://github.com/SimonJF/links/commit/fe036509843a056053c6bd4c1a753b391caeae85

The issue was with trying to calculate the type for bodies of `for` comprehensions generated as the result of shredding, where the body is `Concat ([])`. One solution could be to take the type of the type of the first generator in the query for example, but I'm not entirely sure of the intricacies of this. My "solution" simply checks to see whether a query body is empty, and if so, does not perform compilation but instead simply returns an empty list.

I don't think that's a candidate for merging. In the absence of a fully-typed IR (and as a result, being able to pluck the query type from there for use in shredding), I wonder whether the problem could be solved during query normalisation.

**Comment by jamescheney** [comment]
Thanks. I've asked @wricciot to have a look while re-familiarizing with the query code.

**Comment by slindley** [comment]
RTTI is not a good solution.

This case is easy to fix, though. With some minor tweaks to the code we can normalise before attempting to compute the type. The normal form we obtain is the empty list, which means there is no need to compute the type - or indeed do anything.

There are other instances of `assert false` and `type_of_expression` in the query compilation code that may need checking to ensure that other examples don't trigger this kind of issue.

**Comment by jamescheney** [comment]
@slindley, I'm now a little confused.  `Query.type_of_expression` is already called on `Query.t` expressions, which are produced by simultaneously normalizing / translating from the IR.  So we are already normalizing before attempting to compute the type.

For flat queries, we can presumably just look at the normal form; if it is `Concat([])` then just return an empty result immediately, i.e. do what Simon does in SimonJF@fe03650.  However, I don't see how that approach would fix the problem Simon is running into with deeply nested empty results in shredding.

In fact I'm still not exactly sure how this can possibly be happening "dynamically" (where the `query{}` expression doesn't contain an empty list literal).  In Simon's example the type of the `references`-subquery is perfectly well-defined.  Having a small example that illustrates the problem would definitely help, I've tried to replicate this using a small toy example (similar structure but much simpler) and I haven't managed to so far.  

@SimonJF, it may help if you could enable debugging and copy-paste in what you see when you encounter this error.

**Comment by slindley** [comment]
@jcheney, yes I slightly misread the code in `evalQuery`. It calls `type_of_expression` before `ordered_query` not `eval`. Still, the fix should be easy.

It looks like @SimonJF's example may be relying on shredding which will require a little more care to resolve. (The snippet is rather useless on its own because it doesn't make clear whether the return value is nested or not.)

**Comment by jamescheney** [comment]
Indeed.  I believe it is a nested query result, the full query is here:

https://github.com/SimonJF/skye-gtopdb/blob/7ad36b25368144d50fd28121bf13710a2acc480b/src/diseaseDisplay.links#L130

though there are several other examples illustrating similar problems.

**Comment by SimonJF** [comment]
Yes, it does rely on shredding. Here is the full query (the `skye-gtopdb` repository is currently private):

```
  fun getPathoTargets() {
    query {
      for (pp <-- Tables.pathophysiology)
      where (pp.disease_id == diseaseId)
        for (o <-- Tables.object)
        where (o.object_id == pp.object_id)
        [(id=o.object_id, name=o.name, comments=pp.comments,
          role=pp.role, drugComments=pp.drugs, sideEffects=pp.side_effects,
          therapeuticUse=pp.use,
          references =
            for (pp_ref <-- Tables.pathophysiology_refs)
              where (pp_ref.pathophysiology_id == pp.pathophysiology_id)
              [pp_ref.reference_id])]
    }
```

**Comment by SimonJF** [comment]
I've spent about half an hour trying to mimic the structure and minimise, but I'm having real trouble reproducing the issue with a minimal example, too. I'll undo my hacks and get some in-situ debug information. 

(This is beginning to become more troublesome with the larger DB pages in GtoPdb now, so I'm going to devote some time at least getting the information that can be used to fix this issue).

I would imagine the only way this can happen dynamically is if there is enough information to normalise the query to returning an empty result, or if enough information is deduced during shredding to work out that a query body must be empty  -- but the query is still generated.

**Comment by jamescheney** [comment]
I've looked fairly closely at the shredding code and don't see anywhere we're calling `type_of_expression` and cannot see how it can ever be called during shredding after the (sole) call in `EvalNestedQueries.compile`, even during the later call to `Q.query`.  On the other hand, I don't understand / no longer remember how query generation works in shredding.  

I also can't see how we could determine that the result of the `references` subquery is empty without looking at the database.  So having a debug trace would be very helpful in case the actual problem is something else.

**Comment by SimonJF** [comment]
Okay, I haven't managed to pinpoint this any further, but here is the query code and a debug trace showing each query which is compiled:

https://pastebin.com/B6VVbyN8

The only way I can immediately see the empty query creeping in is if the `false` block of the `if` (statically initialised to `Concat []`) ends up being floated via the normaliser?

```
let compile_shredded : Value.env -> (int * int) option * Ir.computation
                       -> (Value.database * (string * Shred.flat_type) Shred.package) option =
  fun env (range, e) ->
    let v = Q.Eval.eval env e in
    Debug.print ("Query: " ^ (Q.Lang.show v));
   ...
```

I think this is likely, in fact; consider the following application:

```
var db = database "test";
var foo = table "foo" with (foo_id: Int, foo_val: Int) from db;
var bar = table "bar" with (bar_id: Int, bar_val: Int) from db;
var foo2bar = table "foo2bar" with (foo_id: Int, bar_id: Int) from db;
var baz = table "baz" with (baz_id: Int, baz_val: Int) from db;
var foo2baz = table "foo2baz" with (foo_id: Int, baz_id: Int) from db;

typename Model = [(
  fooVal: Int,
  bars: [(barVal: Int, bazIds: [Int])]
)];

sig getModel : (Int) ~%~> Model
fun getModel(fooId) {
  query {
    for (f <-- foo)
    where (f.foo_id == fooId)
      [(fooVal = f.foo_val,
        bars =
          for (f2b <-- foo2bar)
          where (f2b.foo_id == f.foo_id)
            for (b <-- bar)
              where (b.bar_id == f2b.bar_id)
              [(barVal = b.bar_val,
                bazIds =
                  for (f2bz <-- foo2baz)
                  where (false)
                  [f2bz.baz_id])
            ]
      )]
  }
}

getModel(2)
```

This ends up exhibiting the same bug due to the `where (false)`.

**Comment by jamescheney** [comment]
OK.  That's helpful.  I strongly suspect there's at least one other bug, since it should not be the case that shredding can determine the query result is empty without accessing the database (you can see the reference= subquery correctly represented as a "For" earlier).

So, either somehow the query is being executed in nested loop fashion and the inner loops that generate empty queries are being specialized using the results to empty "return" clauses, or there's a bug somewhere that generates empty "return" clauses in this specific setting.

I can't see where in Links codebase the "Query:" debug messages are being printed - is this something you added temporarily?  Where?

Anyway, thanks for the small example - I can duplicate the assert failure now.

**Comment by jamescheney** [comment]
Ah wait, I see where you said where you added the Debug.print.

OK, I just tried running the following toy program, which contains just the above query and the table declarations needed for it:

```
var db = database "gtopdb";
var pathophysiology = table "pathophysiology" with (pathophysiology_id: Int, object_id: Int, disease: String, role: String, drugs: String, side_effects: String, use: String, omim: String, comments: String, orphanet: String, disease_id: Int) from db;
var object = table "object" with (object_id: Int, name: String, last_modified: Int, comments: String, structural_info_comments: String, old_object_id: Int, annotation_status: Int, only_iuphar: Bool, grac_comments: String, only_grac: Bool, no_contributor_list: Bool, abbreviation: String, systematic_name: String, quaternary_structure_comments: String, in_cgtp: Bool, in_gtip: Bool, gtip_comment: String, in_gtmp: Bool, gtmp_comment: String) from db;
var pathophysiology_refs = table "pathophysiology_refs" with (pathophysiology_id: Int, reference_id: Int) from db;

fun foo(diseaseId) {
  query {
      for (pp <-- pathophysiology)
      where (pp.disease_id == diseaseId)
        for (o <-- object)
        where (o.object_id == pp.object_id)
        [(id=o.object_id, name=o.name, comments=pp.comments,
          role=pp.role, drugComments=pp.drugs, sideEffects=pp.side_effects,
          therapeuticUse=pp.use,
          references =
            for (pp_ref <-- pathophysiology_refs)
              where (pp_ref.pathophysiology_id == pp.pathophysiology_id)
              [pp_ref.reference_id])]
  }
}

foo(42)
```

It fails, but only *after* shredding completes, i.e. it fails because I don't have a local copy of the GtoPDB database.  (Technically I lied and used a different local database name).  This I think indicates that I am *not* hitting the bug.

@SimonJF, if you run the above code (verbatim), do you still hit the bug?

My config file is:
```
database_driver=postgresql
database_args=...
relax_query_type_constraint=on
shredding=on
debug=on
```
@SimonJF, can you just confirm / show the equivalent lines of your config?  The reason I ask is that the "inlining/specialization" behavior you're seeing is what I would expect to see if Links attempted to run nested-loop queries using flat query evaluation.  So having a clear confirmation that you're definitely running with shredding enabled would help.

**Comment by slindley** [comment]
Why not at least minimise in situ?

There's loads of junk in your original query which I very much doubt plays a role in the problem you're trying to track down. For instance, I doubt that your result needs to return any fields other than `references` and I'd at least want to know that the three levels of nesting are necessary to exhibit the bug.

I agree with @jamescheney that query normalisation certainly shouldn't reduce a dynamic query to the empty list - because it must happen before accessing the database! So this suggests there may be another bug involved other than the obvious one.

I do wonder whether morally we ought to be extracting something a little more restricted than an actual type from normalised queries. Really what we need is exactly enough information to extract the values. If we know that some substructure of the query is empty then there's no need to know how to extract values from that substructure. We could simulate the kind of thing I have in mind by marking appropriate subformulas of the type using some distinguished type such as an empty type.

**Comment by jamescheney** [comment]
OK, looking at https://pastebin.com/B6VVbyN8 I have a few observations/questions:

1. The `rawInfo` query is a single, large, nested query that should be handled by a single `EvalNestedQuery.compile_shredded` call.  It has (by my count) seven occurrences of collection types in its return type.
2. There are three "Query:" debugging printouts.
- A query that starts out iterating over table "disease".  This seems to correspond to `rawInfo`, it is followed by eight db#exec calls.  The query AST looks like a reasonable one for the normal form of rawInfo, function calls are inlined and "references" fields appear correct.
- A query that starts out iterating over table "interaction".  I don't see anything corresponding to this in the initial Links code snippet.
- Finally, a query that starts out iterating over table `reference`.  I also don't see anything corresponding to this in the initial Links code snippet: there is no (sub)query that iterates over this table there.  (The two `reference` fields in the subqueries actually iterate over the tables `Tables.pathophysiology_refs` and `Tables.immuno_disease2object_refs`
3.  So, it looks like the trace shows the `rawInfo` query succeeding, then another query succeeding, then a third query failing.

Is the Links code snippet at the top of the copy-paste *all* of the Links code that could have contributed to the debug trace, or just the part that seemed relevant?  In particular, I now strongly suspect that there is no problem at all with the `rawInfo` query and that instead the problem is somewhere else with a query over the `reference` table whose result normalizes to something empty, and we may just have been getting the table name `reference` confused with the field name(s).

**Comment by jamescheney** [comment]
So for example if this is the Links source code that generates the `reference` query:
```
fun queryReferences(refIds) {
  var containsReference = contains(refIds);
  query {
  for (ref <-- Tables.reference)
    where (containsReference(ref.reference_id))
      [
        (makeReference(ref.reference_id,
             ref.type, ref.title, ref.article_title, ref.year,
             ref.volume, ref.issue, ref.pages, ref.publisher,
             ref.publisher_address, ref.editors, ref.pubmed_id,
             ref.isbn, ref.authors, false))
      ]
    }
}
```

then what may be happening is:
- `contains(refIds)` gets evaluated to a server-side `False` 
- `False` gets inlined into the query and normalized
- the normal form is `For ... Concat[]` which we know doesn't work.

A workaround (for now) would be to test the special case `refIds == []` outside the query.  It looks like the basic problem is just the one we've already identified, not a subtlety involving shredding.  I only see one other query of the `Tables.references` table and it is nested inside a nested query so probably isn't responsible for the query that's generating an error.

@SimonJF does this help?

**Comment by SimonJF** [comment]
Thanks for the detective work. I was indeed wrong about the normalisation; @jamescheney's diagnosis (and fix) does the job perfectly. Thanks!

It seems in that case that we only need to consider the cases where we either have a static empty list in the query, or can normalise to an empty list.

**Comment by jamescheney** [comment]
Yeah, I thought about this a little more yesterday.  Basically the problem is the current architecture of how query evaluation is split between `evalir`, `query`, `sql` and `eval[Nested]Query` means that there is currently not a good place to fix this (as illustrated in your workaround patch).  It's easy enough to check emptiness for flat queries (as you did) but for nested queries we should be detecting empty-list subexpressions in normal forms of nested queries, and handling them specially.    Right now, the interface between `evalir` and `eval[Nested]Query` obliges the latter to return SQL query strings, including in cases where there isn't a (straightforward) SQL equivalent such as for an empty result.

I think a refactor of the query code is in order, probably to separate out the final SQL generation stage so that before we try to generate SQL we have a chance to look at the normal forms and handle the empty-result cases that have no direct SQL equivalent directly.  I'll discuss how to proceed with @wricciot today.

**Comment by jamescheney** [comment]
I thought about that but wasn't sure that a query without a `from` clause was portable/valid SQL... should check.

I just tried that combined with returning the type `[()]` for `Concat []` in `type_of_expression`.  With shredding disabled, I get an error:
```
links> query {for (x <-- factorials) where (true) []};
*** Error: "Assert_failure core/query/evalQuery.ml:284:16"
links> query {for (x <-- factorials) where (x.i == x.f) []};
*** Error: "Assert_failure core/query/evalQuery.ml:284:16"
```
With shredding enabled, I get different errors:
```
*** Fatal error : Column ?column? had no type info in query's type spec: 1@1:Int, 1@2:Int
```
for example for a query like this:
```
for (x <-- factorials) [(a=x.i, b=for(y <-- factorials) where (false) [(a=1)])]
```
or
```
for (x <-- factorials) [(a=x.i, b=for(y <-- factorials) where (x.i == y.i) [])]
```

So there may be some unhandled corner cases in `eval[Nested]Query` still to consider.

**Comment by slindley** [comment]
Well of course we have to deal with those cases too, but the compiler tells you exactly where to look. For instance we can replace the `assert false` at `evalQuery.ml:284` by `0`.

**Comment by wricciot** [comment]
Hi,

Sorry to be late to the party. To make up for the wasted time, I spent this morning trying to make out where the code fails and to understand your numerous remarks.
I think a very concise (hopefully not imprecise) statement to describe the problem is: we refuse to give a type to normalized queries (`Query.t`) that were originally well-typed. I consider this to be conceptually wrong and I have a gut feeling that any workaround that doesn't assign a type to `Concat []` will be fragile.

I see that James has done some experimenting with typing `Concat [] : [()]`, and Sam seems to be hopeful the rest of the code can be adjusted to accomodate this change. Do I understand correctly that this would fix the bug at the root? 
Of course this only works if we know that, in normal forms, the element type of empty collections is irrelevant and can be replaced by a dummy (this may be obvious to you, but isn't to me...).

James's suggestion to further refactor the query code seems to be orthogonal to me and perhaps should be addressed with #540.

**Comment by jamescheney** [comment]
I was assuming that `assert false` meant "this should never happen and there is no sensible result in this case", not "we hope this will never happen so haven't thought about what a sensible result should be".  Changing the `evalQuery` assert false to 0 then results in this error:
```
links> query {for (x <-- factorials) where (x.i == x.f) []};
*** Fatal error : Column ?column? had no type info in query's type spec: 
```
and I haven't been able to figure out how the `?column?` field name got initialized or what needs to change here yet.

**Comment by jamescheney** [comment]
@wricciot I think the issue is that the query expressions do have types, but if (and only if?) the result of a list-valued subexpression is the empty list, there are infinitely many types.  Choosing `[()]` may be a workaround but there also seem to be many corner cases in query normalization/execution that so far haven't been exercised when this is the case (see also #567).

**Comment by slindley** [comment]
@jamescheney that's the error you got before, presumably because you're still enabling shredding. The `?column?` is the dummy column generated by `select 42` and it doesn't appear in the type. Actually `select where false` seems to work just as well in postgres. Of course `assert false` should mean "this should never happen", but OCaml has a fairly weak type system so it's easy to miss cases.

@wricciot we cannot in general know the exact type of queries beforehand because of polymorphism.

**Comment by jamescheney** [comment]
OK, thanks.  The error above was with shredding disabled, but this makes sense because we would get an empty query result with "?column?" in either case.  So we can hopefully just work around this (for now) by 

In most cases, the type of the query expression / empty list is not polymorphic, and it should always be the case that at run time the type has been instantiated to a ground record type, it's just that we don't currently have enough information at run time to recover it.

I've discussed with @wricciot and will commit my work so far in a PR that he can work on further.  I still tend to think that it would be better to tag the query expression with its inferred (possibly polymorphic) type and use the (hopefully) ground/instantiated version of it in evalir.  Of course this won't work if the IR type information is inconsistent, and we might want to do this in a more sopisticated way than explicit run-time type information (ala Rep/Typeable in GHC).  We will need something like this anyway to implement something like @fehrenbach's type-directed translations.   But this is a medium-term design goal and I'm happy with a workaround for now until we have a clear design.

**Comment by wricciot** [comment]
@slindley I understand that, my point was about the structure of the code: if `type_of_expression` is a partial function, then any code that calls into that must satisfy an invariant about the argument being in the domain of the function: not nice IMHO.

**Comment by wricciot** [comment]
How boring. The example fails with shredding enabled because it ultimately creates the dummy query `select 42 as "@unit@" where false`, expecting it to be compatible with the flattened shredded type for lists of empty records (a list of indices). Then `build_stitch_map` gets confused because it cannot match column names.

I can also see a secondary bug in `build_stitch_map` based on the fact that `unflatten_type` isn't an inverse of `flatten_type` when nested empty records are involved. Anyway I hope to fix the bug tomorrow.

**Comment by jamescheney** [comment]
Fixed by #568
